File: ts/src/cmd/log.ts
import type { RadStore } from '../store';

export interface LogOptions {
  participant?: string;
  file?: string;
  status?: string;
}

export function runLog(store: RadStore, opts: LogOptions): string {
  let operations = store.loadOplog();

  // Sort by timestamp
  operations.sort((a, b) => a.timestamp - b.timestamp);

  // Apply filters
  if (opts.participant) {
    operations = operations.filter(o => o.participantId === opts.participant);
  }

  if (opts.file) {
    operations = operations.filter(o => o.regionId.startsWith(opts.file));
  }

  if (opts.status) {
    const statusLower = opts.status.toLowerCase();
    operations = operations.filter(o => o.status.toLowerCase() === statusLower);
  }

  // Build output
  let output = '';
  for (const op of operations) {
    // Extract content preview (first 50 chars, escape newlines)
    const contentClean = op.content.replace(/\n/g, '\\n').replace(/\r/g, '');
    const contentPreview = contentClean.length > 50
      ? `"${contentClean.slice(0, 47)}..."`
      : `"${contentClean}"`;

    output += `${op.id} [${op.status}]  ${op.participantId}  ${op.regionId}  ${contentPreview}\n`;
  }

  return output;
}

File: rust/cli/src/cmd/log.rs
use crate::store::RadStore;

pub struct LogOptions {
    pub participant: Option<String>,
    pub file: Option<String>,
    pub status: Option<String>,
}

pub fn run_log(store: &RadStore, opts: &LogOptions) -> Result<String, String> {
    let oplog = store.load_oplog()?;
    let mut ops = oplog.get_all_operations();

    // Sort by timestamp
    ops.sort_by_key(|o| o.timestamp);

    // Apply filters
    if let Some(ref p) = opts.participant {
        ops.retain(|o| &o.participant_id == p);
    }

    if let Some(ref f) = opts.file {
        ops.retain(|o| o.region_id.starts_with(f));
    }

    if let Some(ref s) = opts.status {
        let status_lower = s.to_lowercase();
        ops.retain(|o| {
            format!("{:?}", o.status).to_lowercase() == status_lower
        });
    }

    // Build output
    let mut output = String::new();
    for op in ops {
        // Extract content preview (first 50 chars, escape newlines)
        let content_clean = op.content.replace('\n', "\\n").replace('\r', "");
        let content_preview = if content_clean.len() > 50 {
            format!("\"{}...\"", &content_clean[..47])
        } else {
            format!("\"{}\"", content_clean)
        };

        let status_str = format!("{:?}", op.status).to_lowercase();

        output.push_str(&format!("{} [{}]  {}  {}  {}\n",
            op.id,
            status_str,
            op.participant_id,
            op.region_id,
            content_preview));
    }

    Ok(output)
}
